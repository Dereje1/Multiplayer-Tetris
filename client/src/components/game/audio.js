import React from 'react';
import lineClearSoundFile from './styles/clearedline.wav';
import maxLineClearSoundFile from './styles/maxclearedlines.wav';
import looserSoundFile from './styles/Looser.wav';
import winnerSoundFile from './styles/Winner.wav';

export const audioTypes = {
  winnerAudio: winnerSoundFile,
  looserAudio: looserSoundFile,
  clearAudio: lineClearSoundFile,
  maxClearAudio: maxLineClearSoundFile,
};

export const Audio = ({ ...props }) => (
  Object.keys(audioTypes).map(audio => (
    <audio ref={props[audio]} key={audio} src={audioTypes[audio]}>
      <track kind="captions" />
    </audio>
  ))
);
