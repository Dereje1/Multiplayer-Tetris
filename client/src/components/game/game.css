
.democontainer{
  display:flex;
  justify-content: center;
  margin:auto;
}


  