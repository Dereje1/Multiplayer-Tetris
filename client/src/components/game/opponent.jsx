import React from 'react';
import PropTypes from 'prop-types';
import './opponentdescription.css';
// connect to redux
import { connect } from 'react-redux';

import {
  SIMULATE_GAMEPLAY,
} from '../../constants';

import { socket } from '../../Actions/socket';
import { clearCanvas, drawRuble, drawBoundary, drawCells } from './scripts/canvas';

// custom components
import OpponentDescription from './opponentInfo';

// reads from store
const mapStateToProps = state => state;

// writes to store
class Opponent extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      gameState: {},
      status: ['', 0], // [status, data]
      playerPool: [],
      selfSocketId: '',
      opponent: {},
      levelsRaised: 0,
    };
    this.canvasOpponent = React.createRef();

    socket.on('CURRENT_POOL', pool => this.processPool(pool));
    socket.on('INVITATION_RECEIVED', invitedBy => this.processInvite(invitedBy));
    socket.on('GAME_DISCONNECTED', sId => this.disconnectGame(sId));
    socket.on('START_GAME', opp => this.processGameStart(opp));
    socket.on(SIMULATE_GAMEPLAY, oppGame => this.processGame(oppGame));
    socket.on('GAME_END', win => this.processGameEnd(win));
  }
  componentDidMount() {
    this.countGameover = 0;
    socket.emit('PLAYER_JOINED', JSON.stringify(this.props.user));
    console.log('Opponent Mounted!!');
  }

  componentDidUpdate(prevProps, prevState) {
    if (Object.keys(prevState.gameState).length) {
      if (prevState.gameState.points.totalLinesCleared !==
         this.state.gameState.points.totalLinesCleared) {
        this.processFloorRaise();
      }
    }
    if (prevState.status[0] !== this.state.status[0]) {
      if ((this.state.status[0] === 'Playing') || (this.state.status[0] === 'GameOver')) {
        this.props.onDisableExit(true);
      }
    }
  }

  componentWillUnmount() {
    socket.emit('COMPONENT_UNMOUNTED', 'opponent');
    socket.emit('disconnect', '');
  }

  setGame = () => {
    if (Object.keys(this.state.gameState).length) {
      // full deep copy of game state needed as object mutation becomes a problem
      const copyOfState = JSON.parse(JSON.stringify(this.state.gameState));
      if (this.state.status[0] === 'GameOver') {
        clearCanvas(this.canvasOpponentContext, copyOfState);
        return;
      }
      const canvasOpponent = this.canvasOpponent.current;
      canvasOpponent.style.backgroundColor = 'black';
      this.canvasOpponentContext = canvasOpponent.getContext('2d');
      copyOfState.activeShape.unitBlockSize /= 2;
      clearCanvas(this.canvasOpponentContext, copyOfState);
      drawBoundary(this.canvasOpponentContext, copyOfState);
      drawCells(this.canvasOpponentContext, copyOfState.activeShape, true);
      drawRuble(this.canvasOpponentContext, copyOfState, true);
    }
  };

  setUp = () => {
    const opponentsAvailable = !!this.state.playerPool.length;
    if (!opponentsAvailable) {
      this.setState({
        status: ['noopponents', 0],
      });
    } else {
      this.setState({
        status: ['opponents', this.state.playerPool.length],
      });
    }
  }
  setDifficulty = (val) => {
    this.props.onSetDifficulty(val);
  }

  processFloorRaise = () => {
    /*
    Difficulty                                 Description
    -----------------------------------------------------------------------------------
      1               After player clears 4 rows , floor is raised by 1 row on opponent
      2               After player clears 3 rows , floor is raised by 1 row on opponent
      3               After player clears 2 rows , floor is raised by 1 row on opponent
      4               After player clears 1 row  , floor is raised by 1 row on opponent
    */
    const { totalLinesCleared } = this.state.gameState.points;
    const difficultyMap = [[4, 1], [3, 2], [2, 3], [1, 4]]; // [[level, ]]
    const amountNeededForRaise = difficultyMap.filter(d => d[0] === this.props.difficulty)[0][1];
    const targetRaised = Math.floor(totalLinesCleared * (1 / amountNeededForRaise));
    const toRaise = targetRaised - this.state.levelsRaised;
    if (toRaise > 0) this.props.onFloorRaise(Number(toRaise));
    this.setState({
      levelsRaised: this.state.levelsRaised + toRaise,
    });
  }

  /* process socket-in-coming below */
  processPool = (poolData) => {
    const checkSelfSocketId = this.state.selfSocketId === '' ? poolData.self : this.state.selfSocketId;
    const nonSelfPoolData = poolData.pool.filter(p => p.socketId !== checkSelfSocketId);
    const playerChoices = [];
    nonSelfPoolData.forEach((sock, idx) => {
      if (idx < 5) playerChoices.push(sock);
    });

    this.setState({
      playerPool: playerChoices,
      selfSocketId: this.state.selfSocketId === '' ? poolData.self : this.state.selfSocketId,
    }, () => this.setUp());
  }

  disconnectGame = () => {
    console.log('A signal has come in that your Opponent Disconnected the Game!!');
    // add penalty in database for this in the future
    this.props.onGameOver(null);
  }

  processInvite = (host) => {
    const fullPlayerInfo = this.state.playerPool.filter(p => p.socketId === host.hostSocketId)[0];
    this.setState({
      status: ['Invite', [fullPlayerInfo, host.difficulty]],
    });
  }

  processGameStart = (opp) => {
    this.setState({
      opponent: opp,
    });
    let startCounter = 15;
    const gameStartId = setInterval(() => {
      this.setState({
        status: ['PreGame', startCounter],
      });
      startCounter -= 1;
      if (startCounter <= 0) {
        this.setState({
          status: ['Playing', null],
        });
        this.props.onGameEmit({ self: this.state.selfSocketId, opponnent: opp });
        this.props.onReset();
        clearInterval(gameStartId);
      }
    }, 1000);
  }

  processGame = (msg) => {
    if (this.state.status[0] === 'GameOver') {
      clearCanvas(this.canvasOpponentContext, this.state.gameState);
      return;
    }
    const copyOfState = JSON.parse(JSON.stringify(this.state));
    copyOfState.gameState = JSON.parse(msg);
    this.setState(copyOfState, () => this.setGame());
  }

  processGameEnd = (isWinner) => {
    this.props.onClearCanvas();
    this.countGameover += 1;
    if (this.countGameover > 1) return;
    // clearCanvas(this.canvasOpponentContext, this.state.gameState);
    // isWinner , true if client won
    const databaseEntry = isWinner ?
      {
        difficulty: this.props.difficulty,
        multiPlayer: true,
        players: [
          {
            name: this.props.user.displayName,
            _id: this.props.user._id,
            score: this.props.game.points.totalLinesCleared * 50,
            winner: isWinner,
          },
          {
            name: this.state.opponent.displayName,
            _id: this.state.opponent._id,
            score: this.state.gameState.points.totalLinesCleared * 50,
            winner: !isWinner,
          },
        ],
      }
      : null;

    let startCounter = 10;
    this.setState({
      status: ['GameOver', isWinner],
    }, () => this.props.onPause(true));
    console.log(`game over called ${this.countGameover}`);
    const gameEndId = setInterval(() => {
      startCounter -= 1;
      if (startCounter <= 0) {
        if (databaseEntry) this.props.onGameOver(databaseEntry);
        else this.props.onGameOver(null);
        this.props.onDisableExit(false);
        clearInterval(gameEndId);
        this.props.onReset(false);
      }
    }, 1000);
  };

  /* process socket-out-going below */
  requestInvite = (p) => {
    socket.emit('INVITATION_SENT', { hostSocketId: p, difficulty: this.props.difficulty });
  }

  acceptInvite = () => {
    this.props.onSetDifficulty(this.state.status[1][1]);
    this.setState({
      status: ['Loading', null],
    });
    socket.emit('INVITATION_ACCEPTED', [this.state.selfSocketId, this.state.status[1][0].socketId]);
  }

  /* done sockets */

  render() {
    return (
      <div className="opponentContainer">
        <OpponentDescription
          opponentState={this.state}
          difficulty={this.props.difficulty}
          setDifficulty={this.setDifficulty}
          requestInvite={sId => this.requestInvite(sId)}
          acceptInvite={() => this.acceptInvite()}
        />
        <canvas
          ref={this.canvasOpponent}
          width={this.props.game.canvas.canvasMajor.width / 2}
          height={this.props.game.canvas.canvasMajor.height / 2}
        />
      </div>
    );
  }

}

Opponent.defaultProps = {
  game: {},
  user: {},
  onFloorRaise: null,
  onReset: null,
  onGameEmit: null,
  onGameOver: null,
  onSetDifficulty: null,
  onClearCanvas: null,
  onPause: null,
  onDisableExit: null,
  difficulty: 2,
};
Opponent.propTypes = {
  game: PropTypes.objectOf(PropTypes.any),
  user: PropTypes.objectOf(PropTypes.any),
  difficulty: PropTypes.number,
  onFloorRaise: PropTypes.func,
  onReset: PropTypes.func,
  onGameEmit: PropTypes.func,
  onGameOver: PropTypes.func,
  onSetDifficulty: PropTypes.func,
  onClearCanvas: PropTypes.func,
  onPause: PropTypes.func,
  onDisableExit: PropTypes.func,
};

export default connect(mapStateToProps)(Opponent);

