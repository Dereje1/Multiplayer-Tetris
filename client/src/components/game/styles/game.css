.democontainer {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  margin: auto;
}

#smallwindow::before {
  display: block;
  margin-left: 30px;
  font-weight: bold;
  font-size: 35px;
  text-align: left;
  content: 'This app is not designed for screens this size please check back later for updates';
}
/*# sourceMappingURL=game.css.map */