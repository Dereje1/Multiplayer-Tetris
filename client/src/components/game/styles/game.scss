#landing{
    background:#5d95dd67;
    height: 'auto';
    display: flex;
    justify-content: center;
    align-items: center;
}
.democontainer{
    display:flex;
    justify-content: center;
    margin:auto;
}
#smallwindow::before{
    display: block;
    margin-left: 30px;
    font-weight: bold;
    font-size: 35px;
    text-align: left;
    content: 'This app is not designed for screens this size please check back later for updates';
}  
  