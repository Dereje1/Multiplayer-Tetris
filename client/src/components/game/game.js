import React from 'react';
import PropTypes from 'prop-types';

import './styles/game.scss';
// connect to redux and get action creators
import { connect } from 'react-redux';
import {
  getFloorRaiseBoundry, GameActions,
} from '../../redux/actions/tetris';
import { Audio, audioTypes } from './audio';
// custom functions and scripts
import boardReset from './scripts/boardreset';
import tetrisShapes from './scripts/shapes';
import shapeLocator from './scripts/locateShape';
import { runCollisionTest } from './scripts/collision';
import {
  clearCanvas, drawNextShape, drawFloor, drawShape,
} from './scripts/canvas';
import drawScreen from './scripts/drawscreen';
import playerMoves from './scripts/player';
import { processMatch, processSinglePlayer } from './scripts/dbinteraction';
// custom react Components
import Controls from '../controls/controls';
import Opponent from '../oponnent/opponent';
// socket
import { socket as socketConstants, game as gameConstants } from '../../constants/index';
import { clientEmitter } from '../../sockethandler';

const {
  clientEmit: {
    GAME_OVER,
  },
} = socketConstants; // the only emit component makes

const {
  PAUSE, LEVEL_UP, SET_NEXT_SHAPE, SCREEN_UPDATE, COLLISION, INITIALIZE_GAME, RAISE_FLOOR,
} = gameConstants;

// reads from store
const mapStateToProps = state => state;

const actionCreators = {
  GameActions,
};
export class Game extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      multiPlayer: false, // True if in multiplayer mode
      inGameToggle: false, // disallows unmounting of a game in progress in multiplayer
      difficulty: 2, // level transmitted to a guest opponent upon invitation
      floorsRaised: 0, // captures floors raised on opponent
      buttonPause: true, // single player only
      updateFloor: false, // builds floor on next tick if true
      canvasLoaded: false, // loads only once per mount
      windowTooSmall: null,
      lastRefresh: 0, // animation last refresh
      requestAnimation: true, // used a switch to turn animation off/on
    };
    this.canvasMajor = React.createRef();
    this.canvasMinor = React.createRef();
    Object.keys(audioTypes).forEach(audio => (this[audio] = React.createRef()));
  }

  componentDidMount() {
    const { GameActions: Action, socket } = this.props;
    Action(INITIALIZE_GAME, 1, true);
    this.checkWindowSize();
    window.addEventListener('resize', () => this.checkWindowSize());
    if (socket.temp && socket.temp.invitationFrom) this.setState({ multiPlayer: true });
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { lastRefresh } = this.state;
    // component already updates on redux prop change of y-movement for new tick
    // prevent an unnecessary update on last refresh being changed in the state.
    if (nextState.lastRefresh !== lastRefresh) return false;
    return true;
  }

  componentDidUpdate(prevProps) {
    // all optimizations go here
    if (!Object.keys(prevProps.game).length) return;
    const { game: prevGame, socket: prevSocket } = prevProps;
    const { game, socket, GameActions: Action } = this.props;
    const {
      multiPlayer, canvasLoaded, windowTooSmall,
    } = this.state;
    if (windowTooSmall) return;

    /* load Canvas */
    if (!canvasLoaded && this.canvasMajor) this.loadCanvas();

    /* spped up on level up */
    if ((game.points.level > prevGame.points.level)
          && (game.timerInterval > 100)
          && (!multiPlayer)
    ) {
      Action(LEVEL_UP, 50);
    }

    /* draws floor or sets state to do so before the next tick */
    if (game.rubble.boundaryCells.length > 10
       && prevGame.rubble.boundaryCells.length !== game.rubble.boundaryCells.length) {
      if (!game.activeShape.cells.length || game.paused) drawFloor(game, this.canvasContextMajor);
      else this.setState({ updateFloor: true });
    }

    /* an Invitation from another client has been accepted */
    if (!multiPlayer && socket.temp) {
      if (prevSocket.temp && !prevSocket.temp.acceptedInvitation
          && socket.temp.acceptedInvitation) {
        this.setState({ multiPlayer: true }, () => this.resetBoard({ reStart: false }));
      }
    }
    /* Opponent has unmounted after accepting invitation but no game started */
    if (prevSocket && prevSocket.temp && !socket.temp) {
      this.setState({ multiPlayer: false }, () => this.resetBoard({ reStart: false }));
    }
  }

  componentWillUnmount() {
    this.endTick('componentWillUnmount');
    window.removeEventListener('resize', () => {});
  }

  audioProps = () => {
    const props = {};
    Object.keys(audioTypes).forEach((audio) => {
      props[audio] = this[audio];
    });
    return props;
  };

  loadCanvas = () => {
    // loads canvas once on game mount
    const canvasMajor = this.canvasMajor.current;
    const canvasMinor = this.canvasMinor.current;
    canvasMajor.focus();
    canvasMajor.style.backgroundColor = 'black';
    canvasMinor.style.backgroundColor = 'black';
    this.canvasContextMajor = canvasMajor.getContext('2d');
    this.canvasContextMinor = canvasMinor.getContext('2d');
    this.canvasMajor.current.focus();
    this.setState({ canvasLoaded: true });
  };

  resetBoard = (config) => {
    const { game, GameActions: Action } = this.props;
    const resetObject = {
      config,
      stateReset: stateItem => this.setState(stateItem),
      redux: { game, gameReset: floorHeight => Action(INITIALIZE_GAME, floorHeight, true) },
      classItems: {
        canvasContextMajor: this.canvasContextMajor,
        canvasContextMinor: this.canvasContextMinor,
        animationId: this.animationId,
        endTick: this.endTick,
        startTick: this.startTick,
      },
    };
    boardReset(resetObject);
  };

  startTick = async (makeNewShape = true) => {
    const { GameActions: Action } = this.props;
    if (makeNewShape) {
      const data = this.newShape();
      // unable to update store wwithout async, not sure why ??
      await Action(SCREEN_UPDATE, data);
      // eslint-disable-next-line react/destructuring-assignment
      drawShape(this.canvasContextMajor, this.props.game);
    }
    // setTimeout so not too immediately start the animation after new shape
    setTimeout(() => {
      this.setState({ requestAnimation: true });
      this.animationId = requestAnimationFrame(this.tick);
    }, 50);
  };

  tick = (timeStamp) => {
    const { lastRefresh, requestAnimation, updateFloor } = this.state;
    const { game } = this.props;
    // console.log(timeStamp);

    console.dir({ lastRefresh, timeStamp, updateFloor })
    if ((timeStamp - lastRefresh) >= game.timerInterval) {
      console.log('howdy')
      if (updateFloor) drawFloor(game, this.canvasContextMajor);
      this.setState({
        lastRefresh: timeStamp,
        updateFloor: false,
      });
      // can not do moveshape() in setstate callback, will slip, need more investigation
      this.moveShape();
    }
    // recursively call tick if animation state is still on,
    // there maybe a little slippage before it turns off
    if (requestAnimation) this.animationId = requestAnimationFrame(this.tick);
  };

  endTick = (sentBy) => {
    console.log('In end tick')
    if (process.env.NODE_ENV === 'development') console.log(sentBy);
    this.setState({ requestAnimation: false });
    cancelAnimationFrame(this.animationId);
  };

  // get the next shape ypos
  positionForecast = () => {
    const { game } = this.props;
    const copyOfActiveShape = Object.assign({}, game.activeShape);
    copyOfActiveShape.yPosition += game.activeShape.unitBlockSize;
    return copyOfActiveShape;
  };

  newShape = () => {
    // draw next shape on minor and send data of current shape to starttick()
    const { game, GameActions: Action } = this.props;
    const { randomShape, newShapeName, nextShapeInfo } = tetrisShapes.createNewShape(game);
    Action(SET_NEXT_SHAPE, newShapeName);
    drawNextShape(this.canvasContextMinor, nextShapeInfo, game);
    // prepare activeshape data to send to starttick
    [randomShape.boundingBox, randomShape.absoluteVertices] = tetrisShapes.getDims(randomShape);
    const locatedShape = shapeLocator(
      this.canvasContextMajor,
      game.canvas.canvasMajor.width,
      game.canvas.canvasMajor.height,
      randomShape, false,
    );
    const data = {
      activeShape: locatedShape,
      rubble: game.rubble,
    };
    return data;
  };

  moveShape = (newPosition = this.positionForecast()) => {
    const { game, GameActions: Action } = this.props;
    drawScreen(
      {
        updatedShape: newPosition,
        canvasContextMajor: this.canvasContextMajor,
        endTick: this.endTick,
        startTick: this.startTick,
        gameOver: this.gameOver,
        redux: {
          game,
          collide: data => Action(COLLISION, data),
          updateScreen: data => Action(SCREEN_UPDATE, data),
        },
        audio: {
          lineCleared: () => this.clearAudio.current.play(),
          maxLinesCleared: () => this.maxClearAudio.current.play(),
        },
      },
    );
  };

  /* Handle Player Events Below */
  handlePause = () => {
    const { buttonPause } = this.state;
    const { GameActions: Action, game } = this.props;
    this.setState(prevState => ({ buttonPause: !prevState.buttonPause }));
    this.canvasMajor.current.focus();
    Action(PAUSE, !buttonPause);
    if (game.paused) {
      // test if a new game or within a game
      if (game.activeShape.cells.length) this.startTick(false);
      else this.resetBoard({});
    } else this.endTick('Manual Pause');
  };

  floorRaise = (f) => {
    const { game, GameActions: Action } = this.props;
    // Locate Shape on screen and then set .cell prop of activeShape
    const locatedShape = shapeLocator(
      this.canvasContextMajor,
      game.canvas.canvasMajor.width,
      game.canvas.canvasMajor.height,
      this.positionForecast(),
      false,
    );
    const newFloor = getFloorRaiseBoundry(game.rubble, f);
    const collisionResult = runCollisionTest(game, locatedShape, newFloor);
    this.canvasMajor.current.focus();
    if (collisionResult) {
      // right now can not raise floor and collide simultaneously
      console.log('Unable to move floor', collisionResult);
    } else {
      Action(RAISE_FLOOR, game.rubble, { raiseBy: f });
    }
  };

  gamePlay = (e) => {
    const { game } = this.props;
    const ans = (playerMoves(e, game, this.canvasContextMajor));
    if (ans) {
      if (ans === 'forcedown') {
        this.endTick('Down Key');
        this.moveShape();
      } else {
        this.moveShape(ans);
      }
    }
  };

  arrowKeyLag = e => (e.keyCode === 40 ? this.startTick(false) : null);

  /* opponent component Callbacks */
  handleMultiplayer = () => {
    const { user } = this.props;
    const { multiPlayer } = this.state;

    if (user.profile.authenticated) {
      clearCanvas(this.canvasContextMajor, 'All', 'Multi');
      clearCanvas(this.canvasContextMinor, 'All', 'Multi');
      this.setState({
        multiPlayer: !multiPlayer,
      }, () => this.resetBoard({ reStart: false }));
    }
  };

  gameOver = (opponentInfo = null) => {
    const { multiPlayer } = this.state;
    const { game, user, socket } = this.props;
    // Whoever looses first will emit game over while in multiplayer mode
    if (multiPlayer && socket.temp.gameInProgress) {
      clientEmitter(GAME_OVER, socket);
      this.looserAudio.current.play();
    }
    // disregard first local loss signal in multiplayer as another one will come from socket
    if (multiPlayer && !opponentInfo) return;

    const opponent = opponentInfo
      ? processMatch(opponentInfo[0], this.state, this.props, this.winnerAudio)
      : processSinglePlayer({ game, user });

    this.setState({
      buttonPause: true,
    }, () => this.resetBoard({
      reStart: false, keepFloor: false, gameover: true, opponent,
    }));
  };

  checkWindowSize = () => {
    const { multiPlayer } = this.state;
    const minHeight = 700;
    const minWidthSP = 585;
    const minWidthMP = 800;
    if (
      (multiPlayer && window.innerWidth < minWidthMP)
        || (!multiPlayer && window.innerWidth < minWidthSP)
        || (window.innerHeight < minHeight)
    ) this.setState({ windowTooSmall: true });
    else this.setState({ windowTooSmall: null }, () => this.loadCanvas());
  };

  render() {
    const { game, socket } = this.props;
    const {
      difficulty, multiPlayer, inGameToggle,
      buttonPause, floorsRaised, windowTooSmall,
    } = this.state;
    return(
      <div id="landing">
        {
          Object.keys(game).length && !windowTooSmall ?
          <div className="democontainer">
          <Controls
            minorCanvas={this.canvasMinor}
            game={game}
            difficulty={difficulty}
            socketId={socket.mySocketId}
            multiPlayer={[multiPlayer, inGameToggle]}
            pauseButtonState={buttonPause}
            onReset={b => this.resetBoard({ reStart: b })}
            onhandlePause={() => this.handlePause}
            onFloorRaise={() => this.floorRaise(1)}
            onMultiPlayer={() => this.handleMultiplayer}
            allowMultiPlayer={Boolean(Object.keys(socket).length) && socket.usersLoggedIn > 1}
          />
          <canvas
            ref={this.canvasMajor}
            width={game.canvas.canvasMajor.width}
            height={game.canvas.canvasMajor.height}
            tabIndex="0"
            onKeyDown={e => this.gamePlay(e)}
            onKeyUp={e => this.arrowKeyLag(e)}
          />
          {multiPlayer
            ? (
              <Opponent
                onReset={reStart => this.resetBoard(reStart)}
                onFloorRaise={f => this.floorRaise(f)}
                onGameOver={msg => this.gameOver(msg)}
                onCanvasFocus={() => this.canvasMajor.current.focus()}
                onSetDifficulty={d => this.setState({ difficulty: d })}
                toggleMultiplayer={() => this.setState({ inGameToggle: !inGameToggle })}
                difficulty={difficulty}
                floorsRaisedOnOpp={f => this.setState({ floorsRaised: floorsRaised + f })}
              />
            )
            : null
          }
          <Audio {...this.audioProps()} />
        </div>
        :
        <div id="smallwindow" />
        }
      </div>
    )
  }
}

Game.defaultProps = {
  game: {},
  user: {},
  socket: {},
};

Game.propTypes = {
  GameActions: PropTypes.func.isRequired,
  game: PropTypes.objectOf(PropTypes.any),
  user: PropTypes.objectOf(PropTypes.any),
  socket: PropTypes.objectOf(PropTypes.any),
};
export default connect(mapStateToProps, actionCreators)(Game);
