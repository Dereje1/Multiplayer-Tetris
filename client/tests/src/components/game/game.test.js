/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Game } from '../../../../src/components/game/game';

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');
const gameStub = {
    "timerInterval": 700,
    "paused": true,
    "nextShape": "",
    "canvas": {
        "canvasMajor": {
            "width": 300,
            "height": 600
        },
        "canvasMinor": {
            "width": 210,
            "height": 150
        }
    },
    "points": {
        "totalLinesCleared": 0,
        "level": 0,
        "levelUp": 5
    },
    "rubble": {
        "occupiedCells": [],
        "winRows": null,
        "boundaryCells": [
            "0-20",
            "1-20",
            "2-20",
            "3-20",
            "4-20",
            "5-20",
            "6-20",
            "7-20",
            "8-20",
            "9-20"
        ]
    },
    "activeShape": {
        "name": "shapeZ",
        "unitBlockSize": 30,
        "xPosition": 0,
        "yPosition": 0,
        "unitVertices": [],
        "absoluteVertices": [],
        "boundingBox": [],
        "rotationStage": 0,
        "cells": []
    }
}

describe('The Game', () => {
    let props;
    let useRefSpy;
    const focus = jest.fn();
    let counter = 698;
    beforeEach(() => {
        props = {
            GameActions: jest.fn(),
            game: gameStub
        }
        useRefSpy = jest.spyOn(React, "createRef").mockImplementation(() => ({
            current: {
                style: {},
                getContext: () => ({
                    canvas: {
                        width: gameStub.canvas.canvasMajor.width,
                        height: gameStub.canvas.canvasMajor.height
                    },
                    clearRect: jest.fn(),
                    fillRect: jest.fn(),
                    beginPath: jest.fn(),
                    rect: jest.fn(),
                    stroke: jest.fn(),
                    moveTo: jest.fn(),
                    lineTo: jest.fn(),
                    fill: jest.fn()
                }),
                focus
            }
        }))
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => {
            counter++
            if (counter < 702) {
                cb(counter)
            } else {
                return null;
            }
        });
        global.innerWidth = 1000;
        global.innerHeight = 1000;
    })
    afterEach(() => {
        props = null
        useRefSpy = null;
    })
    test('will display message for small screen sizes', () => {
        global.innerWidth = 100;
        global.innerHeight = 100;
        const wrapper = shallow(<Game {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    test('will display the game for allowed screen sizes', () => {
        const wrapper = shallow(<Game {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    test('will start the game', async () => {
        const wrapper = shallow(<Game {...props} />);
        expect(wrapper.state().lastRefresh).toBe(0)
        await wrapper.instance().handlePause()
        jest.advanceTimersByTime(50);
        expect(wrapper.state().lastRefresh).toBe(700)
    });
});