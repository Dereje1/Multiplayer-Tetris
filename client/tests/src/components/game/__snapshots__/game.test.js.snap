// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Game will display message for small screen sizes 1`] = `
<div
  id="landing"
>
  <div
    id="smallwindow"
  />
</div>
`;

exports[`The Game will display the game for allowed screen sizes 1`] = `
<div
  id="landing"
>
  <div
    className="democontainer"
  >
    <Controls
      allowMultiPlayer={false}
      difficulty={2}
      game={
        Object {
          "activeShape": Object {
            "absoluteVertices": Array [],
            "boundingBox": Array [],
            "cells": Array [],
            "name": "shapeZ",
            "rotationStage": 0,
            "unitBlockSize": 30,
            "unitVertices": Array [],
            "xPosition": 0,
            "yPosition": 0,
          },
          "canvas": Object {
            "canvasMajor": Object {
              "height": 600,
              "width": 300,
            },
            "canvasMinor": Object {
              "height": 150,
              "width": 210,
            },
          },
          "nextShape": "",
          "paused": true,
          "points": Object {
            "level": 0,
            "levelUp": 5,
            "totalLinesCleared": 0,
          },
          "rubble": Object {
            "boundaryCells": Array [
              "0-20",
              "1-20",
              "2-20",
              "3-20",
              "4-20",
              "5-20",
              "6-20",
              "7-20",
              "8-20",
              "9-20",
            ],
            "occupiedCells": Array [],
            "winRows": null,
          },
          "timerInterval": 700,
        }
      }
      minorCanvas={
        Object {
          "current": Object {
            "focus": [MockFunction] {
              "calls": Array [
                Array [],
                Array [],
                Array [],
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "getContext": [Function],
            "style": Object {
              "backgroundColor": "black",
            },
          },
        }
      }
      multiPlayer={
        Array [
          false,
          false,
        ]
      }
      onFloorRaise={[Function]}
      onMultiPlayer={[Function]}
      onReset={[Function]}
      onhandlePause={[Function]}
      pauseButtonState={true}
      socketId=""
    />
    <canvas
      height={600}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      tabIndex="0"
      width={300}
    />
    <Audio
      clearAudio={
        Object {
          "current": Object {
            "focus": [MockFunction] {
              "calls": Array [
                Array [],
                Array [],
                Array [],
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "getContext": [Function],
            "style": Object {},
          },
        }
      }
      looserAudio={
        Object {
          "current": Object {
            "focus": [MockFunction] {
              "calls": Array [
                Array [],
                Array [],
                Array [],
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "getContext": [Function],
            "style": Object {},
          },
        }
      }
      maxClearAudio={
        Object {
          "current": Object {
            "focus": [MockFunction] {
              "calls": Array [
                Array [],
                Array [],
                Array [],
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "getContext": [Function],
            "style": Object {},
          },
        }
      }
      winnerAudio={
        Object {
          "current": Object {
            "focus": [MockFunction] {
              "calls": Array [
                Array [],
                Array [],
                Array [],
                Array [],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "getContext": [Function],
            "style": Object {},
          },
        }
      }
    />
  </div>
</div>
`;
